cpp/src/io/parquet/page_decode.cuh

```diff
@@ -1147,16 +1155,20 @@ inline __device__ bool setupLocalPageInfo(page_state_s* const s,
       }
       // Special check for downconversions
       s->dtype_len_in = s->dtype_len;
-      if (s->col.converted_type == DECIMAL && data_type == FIXED_LEN_BYTE_ARRAY) {
-        s->dtype_len = [dtype_len = s->dtype_len]() {
-          if (dtype_len <= sizeof(int32_t)) {
-            return sizeof(int32_t);
-          } else if (dtype_len <= sizeof(int64_t)) {
-            return sizeof(int64_t);
-          } else {
-            return sizeof(__int128_t);
-          }
-        }();
+      if (data_type == FIXED_LEN_BYTE_ARRAY) {
+        if (s->col.converted_type == DECIMAL) {
+          s->dtype_len = [dtype_len = s->dtype_len]() {
+            if (dtype_len <= sizeof(int32_t)) {
+              return sizeof(int32_t);
+            } else if (dtype_len <= sizeof(int64_t)) {
+              return sizeof(int64_t);
+            } else {
+              return sizeof(__int128_t);
+            }
+          }();
+        } else {
+          s->dtype_len = sizeof(string_index_pair);
```

Comment:
This is another case of the type len indicating the type :(.  Look at the comment around line 1225 in the new file...something about "dtype_len is a lie". The real len should be `sizeof(size_type)`, but then we'd think it's a hash. So for strings (and now fixed_len_byte_array) we just leave it as `sizeof(string_index_pair)`, which it would have had to be before the sweeping string decode changes.

Reply:
> Is this the "FIXED_LEN_BYTE_ARRAY but not DECIMAL" case(hard to match block in the diff)?

Yes, we decode data as strings as an intermediate output and eventually the final list will be generated from strings. This is for now the best way to take flat data as input and produce nested output.

Reply:
Thank you for the explanations, now this makes sense.